nomad/secure_variables_endpoint.go

```diff
@@ -1,133 +1,470 @@
+// HTTP API.
 type SecureVariables struct {
 	srv       *Server
 	logger    hclog.Logger
 	encrypter *Encrypter
+
+	// ctx provides context regarding the underlying connection, so we can
+	// perform TLS certificate validation on internal only endpoints.
+	ctx *RPCContext
 }
 
-func (sv *SecureVariables) Create(args *structs.SecureVariablesUpsertRequest, reply *structs.SecureVariablesUpsertResponse) error {
-	if done, err := sv.srv.forward("SecureVariables.Create", args, args, reply); done {
+// Upsert creates or updates secure variables held within Nomad.
+func (sv *SecureVariables) Upsert(
+	args *structs.SecureVariablesUpsertRequest,
+	reply *structs.SecureVariablesUpsertResponse) error {
+
+	if done, err := sv.srv.forward(structs.SecureVariablesUpsertRPCMethod, args, args, reply); done {
 		return err
 	}
+	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "upsert"}, time.Now())
 
-	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "create"}, time.Now())
+	// Perform the ACL token resolution.
+	aclObj, err := sv.srv.ResolveToken(args.AuthToken)
 
-	// TODO: implement real ACL checks
-	if aclObj, err := sv.srv.ResolveToken(args.AuthToken); err != nil {
-		return err
-	} else if aclObj != nil && !aclObj.IsManagement() {
-		return structs.ErrPermissionDenied
+	switch err {
+	case nil:
+		// If ACLs are enabled, ensure the caller has the submit-job namespace
+		// capability.
+		if aclObj != nil {
+			hasSubmitJob := aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilitySubmitJob)
+			if !hasSubmitJob {
+				return structs.ErrPermissionDenied
+			}
+		}
+	default:
+		// This endpoint is generally called by Nomad nodes, so we want to
+		// perform this check, unless the token resolution gave us a terminal
+		// error.
+		if err != structs.ErrTokenNotFound {
+			return err
+		}
+
+		// Attempt to lookup AuthToken as a Node.SecretID and return any error
+		// wrapped along with the original.
+		node, stateErr := sv.srv.fsm.State().NodeBySecretID(nil, args.AuthToken)
+		if stateErr != nil {
+			var mErr multierror.Error
+			mErr.Errors = append(mErr.Errors, err, stateErr)
+			return mErr.ErrorOrNil()
+		}
+
+		// At this point, we do not have a valid ACL token, nor are we being
+		// called, or able to confirm via the state store, by a node.
+		if node == nil {
+			return structs.ErrTokenNotFound
+		}
 	}
 
-	sv.logger.Trace("TODO") // silences structcheck lint
+	// Use a multierror, so we can capture all validation errors and pass this
+	// back so fixing in a single swoop.
+	var mErr multierror.Error
 
-	// TODO: placeholder for serialization and encryption
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
-	err := enc.Encode(args.Data.UnencryptedData)
-	if err != nil {
+	// Iterate the secure variables and validate them. Any error results in the
+	// call failing.
+	for _, i := range args.Data {
+		if err := i.Validate(); err != nil {
```

Comment:
If we've failed validation we should `continue` so that we don't try to encrypt an invalid variable.